### PR TITLE
Feature/53/handle peer disconnection

### PR DESCRIPTION
**Pull request by tharindal**

### CODE REVIEW

File: saito-rust/src/saito/network_controller.rs [hunk 1]
```diff
@@ -85,7 +85,7 @@ impl NetworkController {
         debug!("connecting to peer : {:?}", url);
         let result = connect_async(url.clone()).await;
         if result.is_err() {
-            warn!("failed connecting to peer : {:?}", peer);
+            warn!("failed connecting to peer : {:?}", peer); //TODO : Retry connecting after an interval
```
Comment (SankaD) [inline saito-rust/src/saito/network_controller.rs:88]: you can setup clion to auto format with rustfmt. I think this line will break into 2 after formatting.

**Review by SankaD — CHANGES_REQUESTED** [review]

File: saito-core/src/core/routing_controller.rs [hunk 2]
```diff
@@ -228,8 +231,31 @@ impl RoutingController {
...
+            let peer = result.unwrap();
+
+            if peer.static_peer_config.is_some() {
+                // This means the connection has been initiated from this side, therefore we must
+                // try to re-establish the connection again
+
+                info!("Static peer disconnected, reconnecting .., Peer ID = {}, Public Key = {:?}",
+                    peer.peer_index, hex::encode(peer.peer_public_key));
+
+                self.io_interface.connect_to_peer(peer.static_peer_config.as_ref().unwrap().clone()).await.unwrap();
```
[the remark this answers is not in the document]
Comment (tharindal) [reply saito-core/src/core/routing_controller.rs:247]: Isn't that going to be handled inside the network_controller ? if we are going to retry with timers isn't it the best place to do so?

File: saito-core/src/core/data/peer_collection.rs [hunk 3]
```diff
@@ -16,13 +16,17 @@ impl PeerCollection {
...
             return None;
         }
-        let result = result.unwrap();
-        let result = self.index_to_peers.get(result);
-        result
+
+        return self.find_peer_by_index(*result.unwrap());
+    }
+
+    pub fn find_peer_by_index(&self, peer_index : u64) -> Option<&Peer> {
```
[the remark this answers is not in the document]
Comment (tharindal) [reply saito-core/src/core/data/peer_collection.rs:29]: Is there any advantage here ?

[the remark this answers is not in the document]
Comment (tharindal) [reply saito-core/src/core/data/peer_collection.rs:29]: Ignoring as per your feedback

**Review by SankaD — APPROVED** [review]

**Review by SankaD — APPROVED** [review]

**Pull request merged**